ruby/lib/google/protobuf/ffi/descriptor.rb

```diff
@@ -0,0 +1,765 @@
+          pool = Google::Protobuf::ObjectCache.get(pool_def)
+          raise "Cannot find pool in ObjectCache!" if pool.nil?
+          descriptor = pool.descriptor_class_by_def[msg_def.address]
+          if descriptor.nil?
+            pool.descriptor_class_by_def[msg_def.address] = private_constructor(msg_def, pool)
+          else
+            descriptor
+          end
+        end
+      end
+
+      ##
+      # Great write up of this strategy:
+      # See https://blog.appsignal.com/2018/08/07/ruby-magic-changing-the-way-ruby-creates-objects.html
+      def self.new(*arguments, &block)
+        raise "Descriptor objects may not be created from Ruby."
+      end
+
+      def to_s
+        inspect
+      end
+
+      def inspect
+        "Descriptor - (not the message class) #{name}"
+      end
+
+      def file_descriptor
+        @descriptor_pool.send(:get_file_descriptor, Google::Protobuf::FFI.get_message_file_def(self))
+      end
+
+      def name
+        @name ||= Google::Protobuf::FFI.get_message_fullname(self)
+      end
+
+      def each_oneof &block
+        n = Google::Protobuf::FFI.oneof_count(self)
+        0.upto(n-1) do |i|
+          yield(Google::Protobuf::FFI.get_oneof_by_index(self, i))
+        end
+        nil
+      end
+
+      def each &block
+        n = Google::Protobuf::FFI.field_count(self)
+        0.upto(n-1) do |i|
+          yield(Google::Protobuf::FFI.get_field_by_index(self, i))
+        end
+        nil
+      end
+
+      def lookup(name)
+        Google::Protobuf::FFI.get_field_by_name(self, name, name.size)
+      end
+
+      def lookup_oneof(name)
+        Google::Protobuf::FFI.get_oneof_by_name(self, name, name.size)
+      end
+
+      def msgclass
+        @msg_class ||= build_message_class
+      end
+
+      private
+
+      extend Google::Protobuf::Internal::Convert
+
+      def initialize(msg_def, descriptor_pool)
+        @msg_def = msg_def
+        @msg_class = nil
+        @descriptor_pool = descriptor_pool
+      end
+
+      def self.private_constructor(msg_def, descriptor_pool)
+        instance = allocate
+        instance.send(:initialize, msg_def, descriptor_pool)
+        instance
+      end
+
+      def wrapper?
+        if defined? @wrapper
+          @wrapper
+        else
+          @wrapper = case Google::Protobuf::FFI.get_well_known_type self
+          when :DoubleValue, :FloatValue, :Int64Value, :UInt64Value, :Int32Value, :UInt32Value, :StringValue, :BytesValue, :BoolValue
+            true
+          else
+            false
+          end
+        end
+      end
+
+      def self.get_message(msg, descriptor, arena)
+        return nil if msg.nil? or msg.null?
+        message = ObjectCache.get(msg)
+        if message.nil?
+          message = descriptor.msgclass.send(:private_constructor, arena, msg: msg)
+        end
+        message
+      end
+
+      def pool
+        @descriptor_pool
+      end
+
+      def build_message_class
```

Comment:
I think it would help readability if this method was in its own file called message.rb.

Reply:
Done.